### PR TITLE
Expressions as annotation values

### PR DESCRIPTION
**Pull request by stewsk**
Unrelated: in TOC, changed the order of items so that it matches the order of the corresponding sections in the text below

### CODE REVIEW

**Comment by stewsk** [comment]
@bugwelle Do we have to add the "no check" marker to the examples with the new syntax?

**Comment by bugwelle** [comment]
@stewsk yes. In v4.4, there was a Compiler error which we emitted in the parser already.

**Comment by bugwelle** [comment]
FYI @stewsk : With https://github.com/cap-js/docs/pull/579 , the snippet checker exclusions aren't needed anymore.

File: cds/cdl.md [hunk 1]
```diff
@@ -983,14 +986,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+by CDS parsers or linkers. They're interpreted and evaluated only on consumption-specific modules.
+For example, for SAP Fiori models, it's the _4odata_ and _2edm(x)_ processors.
+:::
+
+
+### Expressions as Annotation Values (beta) {#expressions-as-annotation-values}
+
+::: warning
+Expressions in annotation values are released as beta feature.
+We provide an early preview for the functionality to allow you to experiment with it and provide feedback.
```
Comment (bugwelle) [inline cds/cdl.md:1006]:
```suggestion
We provide an early preview of this functionality to allow you to experiment with it and provide feedback.
```

File: cds/cdl.md [hunk 2]
```diff
@@ -983,14 +986,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+
+#### Name resolution
+
+Each path in the expression is checked. For an annotation assigned to an entity,
+the first path step is resolved as element of the entity. For an annotation assigned
+to an entity element, the first path step is resolved as the annotated element or its siblings.
+If the annotation is assigned to a subelement of a structured element, the top level
+elements of the entity can be accessed via `$self`.
+A parameter `par` of a parametrized entity can be accessed like in a query with `:par`.
+If a path cannot be resolved successfully, compilation fails with an error.
```
Comment (bugwelle) [inline cds/cdl.md:1040]: I feel as if this paragraph should be a bullet list.

File: cds/cdl.md [hunk 3]
```diff
@@ -983,14 +986,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+
+In contrast to `@aReference: foo.bar`, a single reference written as expression `@aRefExpr: ( foo.bar )`
+is checked by the compiler.
+
+#### CSN Representation
+
+In CSN, the expression is represented as a record with two properties:
+* A string representation of the expression is stored in property `=`.
+* A tokenized representation of the expression is stored in one of the properties
+`xpr`, `ref`, `val`, `func`, or `list` (like if the expression was written in a query).
```
Comment (bugwelle) [inline cds/cdl.md:1050]: Do you want to list _all_? That would be:

```
  'ref', 'xpr', 'list', 'literal', 'val',
  '#', 'func', 'args', 'SELECT', 'SET',
  'cast',
```

File: cds/cdl.md [hunk 4]
```diff
@@ -983,14 +986,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+@Foo3 : (a + c)
+entity E {
+  a : Integer;
+  b : Integer;
+  c : Integer;
+}
+
+entity P as projection on E {
+  a,
+  b as x
```
Comment (bugwelle) [inline cds/cdl.md:1093]:
```suggestion
  b as x,
```

File: cds/cdl.md [hunk 5]
```diff
@@ -983,14 +987,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+Currently, paths are not rewritten in propagated annotations. In a projection, for example,
+all elements used in an annotation expression must be projected without renaming.
+Thus, for the time being we recommend to use the feature mainly in top-level projections.
+
+Example:
+<!-- cds-mode: upcoming, cds-compiler v4.5 -->
+```cds
+@Foo1 : (a)
+@Foo2 : (a + b)
+@Foo3 : (a + c)
```
Comment (bugwelle) [inline cds/cdl.md:1084]: Maybe use names that are more "bookshop friendly"? 😄

File: cds/cdl.md [hunk 6]
```diff
@@ -983,14 +987,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+}
+```
+
+All three annotations are propagated to the projection `P`. The propagated `@Foo1` is still valid.
+The propagated annotation `@Foo2` is invalid, as the referenced element `b` has been renamed to `x`.
+The propagated annotation `@Foo3` is invalid, as element `c` of `E` has not been projected at all.
+This results in a compiler error. To make it work, you would have to explicitly overwrite annotations
+`@Foo2` and `@Foo3` at `P`.
+
+The compiler is going to take care of renamings and rewrite references in propagated annotations
```
Comment (bugwelle) [inline cds/cdl.md:1103]:
```suggestion
The compiler is going to take care of renaming and rewriting references in propagated annotations
```

File: cds/cdl.md [hunk 7]
```diff
@@ -983,14 +987,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+in a later release. The CSN representation of propagated annotation expressions may change even
+if today no error is issued. Propagated annotation expressions that today are accepted may lead
+to an error in the future when the implementation is improved.
+
+#### CDS Annotations
+
+Using an expression as annotation value only makes sense if the evaluator of the annotation is
+prepared to deal with the new CSN representation.
+Currently the CAP runtimes only support expressions in the `where` property of the `@restrict` annotation.
```
Comment (bugwelle) [inline cds/cdl.md:1112]:
```suggestion
Currently, the CAP runtimes only support expressions in the `where` property of the `@restrict` annotation.
```

File: cds/cdl.md [hunk 8]
```diff
@@ -983,14 +987,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+<!-- cds-mode: upcoming, cds-compiler v4.5 -->
+```cds
+entity Orders @(restrict: [
+    { grant: 'READ', to: 'Auditor', where: (AuditBy = $user.id) }
+  ]) {/*...*/}
+```
+
+More annotations are going to follow in upcoming releases.
+
+Of course you can use this feature also in your custom annotations, where you control the code that evaluates
```
Comment (bugwelle) [inline cds/cdl.md:1123]:
```suggestion
Of course, you can use this feature also in your custom annotations, where you control the code that evaluates
```

File: cds/cdl.md [hunk 9]
```diff
@@ -983,14 +987,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+the annotations.
+
+#### OData Annotations
+
+The OData backend of CAP doesn't yet support expression valued annotations. This is planned for
+a later release.
+If you use the new expression syntax for OData annotations, the expression and contained references
+are not correctly translated, and the resulting EDMX will change once the OData support is available.
```
Comment (bugwelle) [inline cds/cdl.md:1131]: Well, in fact, you get a compilation error as of now 😄

**Review by bugwelle — APPROVED** [review]
Approved, but please see comments.

File: cds/cdl.md [hunk 10]
```diff
@@ -983,14 +987,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+
+In contrast to `@aReference: foo.bar`, a single reference written as expression `@aRefExpr: ( foo.bar )`
+is checked by the compiler.
+
+#### CSN Representation
+
+In CSN, the expression is represented as a record with two properties:
+* A string representation of the expression is stored in property `=`.
+* A tokenized representation of the expression is stored in one of the properties
+`xpr`, `ref`, `val`, `func`, or `list` (like if the expression was written in a query).
```
Comment (stewsk) [inline cds/cdl.md:1050]:
```suggestion
`xpr`, `ref`, `val`, `func`, etc. (like if the expression was written in a query).
```

File: cds/cdl.md (hunk 3 above)
Comment (stewsk) [reply cds/cdl.md:1050]: No, actually not.

File: cds/cdl.md (hunk 6 above)
Comment (stewsk) [reply cds/cdl.md:1103]: That's not precisely what I want to say ... I'll try again ...

File: cds/cdl.md [hunk 11]
```diff
@@ -983,14 +987,149 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+}
+```
+
+All three annotations are propagated to the projection `P`. The propagated `@Foo1` is still valid.
+The propagated annotation `@Foo2` is invalid, as the referenced element `b` has been renamed to `x`.
+The propagated annotation `@Foo3` is invalid, as element `c` of `E` has not been projected at all.
+This results in a compiler error. To make it work, you would have to explicitly overwrite annotations
+`@Foo2` and `@Foo3` at `P`.
+
+The compiler is going to take care of renamings and rewrite references in propagated annotations
```
Comment (stewsk) [inline cds/cdl.md:1103]:
```suggestion
The compiler is going to take care of renamed elements and rewrites references in propagated annotations
```

File: cds/cdl.md (hunk 9 above)
Comment (stewsk) [reply cds/cdl.md:1131]: Not necessarily .

File: cds/cdl.md (hunk 4 above)
Comment (stewsk) [reply cds/cdl.md:1093]: Don't like the look of that extra "," ...

File: cds/cdl.md (hunk 5 above)
Comment (renejeglinsky) [reply cds/cdl.md:1084]: If that is possible I'd be in favor of it :)

**Review by renejeglinsky — APPROVED** [review]

File: cds/cdl.md [hunk 12]
```diff
@@ -983,14 +987,151 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+as expression `@aValueExpr: ( 11 )`, respectively.
+
+#### Propagation
+
+[Annotations are propagated](#annotation-propagation) in views/projections, in includes, or along type references.
+Currently, paths are not rewritten in propagated annotations. In a projection, for example,
+all elements used in an annotation expression must be projected without renaming.
+Thus, for the time being we recommend to use the feature mainly in top-level projections.
+
+Example:
```
Comment (stewsk) [inline cds/cdl.md:1079]: @bugwelle @renejeglinsky Do you like that one better?

Comment (renejeglinsky) [reply cds/cdl.md:1079]: To be honest, I understand it now better. Thanks :)

Comment (bugwelle) [reply cds/cdl.md:1079]: Better, thanks! :)

File: cds/cdl.md [hunk 13]
```diff
@@ -983,14 +987,151 @@ As described in the [CSN spec](./csn#literals), the previously mentioned annotat
...
+```cds
+@anExpression: ( foo.bar * 11 )
+```
+
+Syntactically, the same expressions are supported as in a select item or in the where clause of a query,
+except subqueries. The expression can of course also be a single reference or a simple value:
+<!-- cds-mode: upcoming, cds-compiler v4.5 -->
+```cds
+@aRefExpr: ( foo.bar )
+@aValueExpr: ( 11 )
```
Comment (agoerler) [inline cds/cdl.md:1023]: Maybe also give examples where the expression is a predicate

**Pull request merged**